app/controllers/concerns/saml_idp_auth_concern.rb

```diff
@@ -140,9 +140,19 @@ def link_identity_from_session_data
       link_identity(
         ial: resolved_authn_context_int_ial,
         rails_session_id: session.id,
+        email_address_id: email_address_id,
       )
   end
 
+  def email_address_id
+    return nil unless IdentityConfig.store.feature_select_email_to_share_enabled
+    return user_session[:selected_email_id] if user_session[:selected_email_id].present?
+    identity = current_user.identities.find_by(service_provider: sp_session['issuer'])
+    email_id = identity&.email_address_id
+    return email_id if email_id.is_a? Integer
+    EmailContext.new(current_user).last_sign_in_email_address.id
```

Comment:
Similar to my comments at https://github.com/18F/identity-idp/pull/10951#discussion_r1719811578, I think this method should _only_ evaluate the session value and `last_sign_in_email_address`, and then we should update `IdentityLinker#link_identity` to only update the identity if the value provided is non-`nil`. That way, we don't have to mess with querying the identity, and don't have to update the identity unless it's changing.

But that's non-blocking, let's create a separate ticket for that.

Reply:
Would this take care of the redundant querying? (it's similar to the change I made for https://github.com/18F/identity-idp/pull/10951#discussion_r1719811578)

def email_address_id
    return nil unless IdentityConfig.store.feature_select_email_to_share_enabled
    return user_session[:selected_email_id] if user_session[:selected_email_id].present?
    identity = current_user.identities.find_by(service_provider: sp_session['issuer'])
    identity&.email_address_id || EmailContext.new(current_user).last_sign_in_email_address.id
  end
  
  Or should that one also be looked at in the same follow on ticket?

Reply:
I guess there's a bit of an open question of: After this feature goes live, what is the expected behavior for someone who signs in to a partner they already consented? Should we save `email_address_id`?

That seems to be what's implemented here, and I think that could make sense, but I'd want LG-13665 to be completed before we enabled the feature in that case, to allow someone to easily change an email address associated with a partner.

Reply:
Ticket: [LG-14354](https://cm-jira.usa.gov/browse/LG-14354)